<a href="https://colab.research.google.com/github/HIT-YL/py-03/blob/master/py03_exam_mid.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 중간필기시험: 파이썬 프로그래밍 (2018.10.15)

## 정답을 설명하는 노우트를 완성해서 각자의 github에 올리시오.

### 업로드 파일명: pynn_exam_mid.ipynb  (pynn은 id)

***

In [1]:
# [1]
# 다음 식의 연산 결과는?

2^4

6

**풀이** : ^는 배타적 XOR 연산을 나타내는 기호다.

In [3]:
# 2^4 계산

# 2 =  10
# 4 = 100
# -------
#     110

# 100 = 6이므로 정답은 6이다.

6

---

In [4]:
# [2]
# 다음 식의 연산 결과는?

2**4

16

**풀이** : **는 제곱을 나타내는 기호다.

In [0]:
# 2**4 계산

# 2의 4승 = 2*2*2*2
# 정답은 16이다.

---

In [7]:
# [3]
# BMI는 키와 체중을 이용하여 계산한 지수로 체중(kg)÷(키(m)**2) 입니다. 
# 즉, 체중(weight)을 키(height)의 제곱으로 나눈 값이 BMI(체질량지수)입니다.
#

height = float(input("Your height in meters:"))
weight = int(input("Your weight in kilograms:"))

# bmi를 소숫점 첫째자리까지 구하는 식은?
# hint: use round()

bmi = round((weight/(height*height)),1)

bmi

Your height in meters:1.61
Your weight in kilograms:39


15.0

**풀이** : round는 소수점 n번째 자리까지를 구하도록 한다. (n+1번째 자리에서 반올림)

---

In [10]:
# [4,5]
# 다음 리스트는 도쿄의 낮 평균 기온 (1950년부터 2000년까지)이댜.

tokyo_temps = [15.1, 15.4, 15.2, 15.4, 17.0, 16.9]

import numpy as np  # [4] 파이썬 모듈을 연결하는 키워드는? 

max(tokyo_temps)

np.argmax(tokyo_temps) #[5] 결과는?  

4

[4] **풀이** : 파이썬 모듈을 연결하는 키워드는 import이다.       
[5] **풀이** : argmax는 최댓값을 갖는 인덱스를 출력하는 명령어이다.

---

In [11]:
# [6,7]
# 영어 단어의 출현 수를 딕셔너리를 이용해서 계산하는 코드이다.
# 문단 line을 구성하는 단어들을 분리하는 정확한 함수식은?

line = """If you can dream it, you can do it. 
- Walt Disney"""

wordcnt = {}     #[6] 딕셔너리를 초기화

for word in line.split():  #[7] 단어들을 분리하는 정확한 함수식은?
    if word in wordcnt:
        wordcnt[word] = wordcnt[word]+1
    else:
        wordcnt[word] = 1
        
print(wordcnt)            # 결과 표시

#[7. 보기] split(',') , split() , split(':') , split('.')



{'If': 1, 'you': 2, 'can': 2, 'dream': 1, 'it,': 1, 'do': 1, 'it.': 1, '-': 1, 'Walt': 1, 'Disney': 1}


[6] **풀이** : 딕셔너리는 {}를 통해 아무것도 입력하지 않음으로써 초기화를 실행한다.           
[7] **풀이** : split()는 공백을 기준으로 단어를 분리하는 함수식이다.

---

In [12]:
# [8]
# 리스트를 set로 변환하기
# 
codon = ['ATG', 'GGC', 'TCC', 'AAG', 'TTC', 
         'TGG', 'GGC', 'GAC', 'TCC']

s_codon = set(codon)      # 리스트를 set으로 변환

print(len(codon), len(s_codon))    #[8] 출력 결과는?

9 7


**풀이** : len(codon)은 codon의 값을 모두 헤아린다. 그러나 len(s_codon)은 s_codon이 condon을 set으로 변환한 것이므로, 중복을 허용하지 않는다는 set의 성격에 따라 중복되는 값인 GGC와 TCC를 한 번씩만 헤아리게 된다.

---

In [15]:
# [9]
# 튜플의 리스트를 이용한 다음 식의 연산 결과는?
# 
tank_data = [("A형 전차", 38, 80, 75), ("B형 전차", 42, 50, 37),
             ("X형 전차", 20, 17, 57), ("Y형 전차", 40, 50, 75),
             ("Z형 전차", 39, 51, 75)]

# 전차의 제원을 더해 반환하는 함수
def evaluate_tankdata(tup):
    return tup[1]+tup[2]+tup[3]

print(evaluate_tankdata(tank_data[3]))  #[9] 출력 결과는?

165


**풀이** : tank_data[3]은 tank_data의 인덱스3에 해당하는 값들 중, 그 값 안의 인덱스1부터 3까지를 더한다.                  
tank_data[3] = ("Y형 전차", 40, 50, 75)이고, 이중에서 인덱스 1,2,3 각각의 값은 40, 50, 75이므로 이를 더하면 165가 된다.

---

In [18]:
# [10]
# 다음 for-loop의 출력 결과는?
#

seq = ["야구", "축구", "농구"]

# range()를 사용한 루프 카운터
for cnt in range(len(seq)):
    print(cnt, seq[cnt], end=',')

0 야구,1 축구,2 농구,

**풀이** : print에 걸린 cnt는 0부터 마지막 인덱스까지 1씩 증가한다. seq[cnt]는 해당하는 인덱스 값을 출력하여 순서대로 야구, 축구, 농구가 출력되며, end에 ','를 넣음으로써 줄바꿈을 ,로 변환하여 표기된다.

---

In [19]:
# [11]
# 다음은 두 개의 리스트를 묶어서 아래와 같이 출력하는 코드이다.
# 두 개의 리스트를 묶는 함수는?

for n, w in zip([1, 2, 3, 4], ['a', 'b', 'c', 'd']):
    print(n, w)
    
# 출력결과    

1 a
2 b
3 c
4 d


**풀이** : 두 개의 리스트를 묶는 함수는 zip이다.

---

In [21]:
# [12]
# 다음은 구글 colab cloud로 로컬 파일들을 업로드하는 과정이다.
# files 객체 UI를 이용해서 업로드한 파일들은 uploaded 변수에 튜플로 저장된다.

from google.colab import files
uploaded = files.upload()

for fn in uploaded.keys():  #[12] 튜플의 함수는?
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

#[12.보기] items(), values(), keys(), map()

Saving foo.txt to foo.txt
User uploaded file "foo.txt" with length 316 bytes


**풀이** : uploaded 뒤에 들어갈 튜플의 함수는 keys()이다.

---

In [22]:
# [13-17]
# 리스트의 인덱싱과 슬라이싱
#

a=[10,20,30,40,50,60]

b=[10,20,30,"HCit",[True,3.14]]


print("a[-5] ==", a[-5]) #[13] 출력 결과는?


print("b[-1][0] ==", b[-1][0]) #[14] 출력 결과는?


print("a[:-3]==",a[:-3]) #[15] 출력 결과는?


print(b[-2][1])  #[16] 출력 결과는?


print(a[::-2])  #[17] 출력 결과는?

a[-5] == 20
b[-1][0] == True
a[:-3]== [10, 20, 30]
C
[60, 40, 20]


[13] **풀이** : a[-5]는 뒤에서 5번째에 해당하는 값을 출력하므로 출력되는 값은 20이다.        
[14] **풀이** : b[-1][0]은 앞에서부터 확인하면 b[-1]=[True, 3.14]이고, b[-1][0]은 b[-1]의 [0]을 출력하는 것이므로 답은 True가 출력된다.            
[15] **풀이** : a[:-3]은 : 앞 공백은 인덱스 첫 값을 가리키는 것이므로 즉 처음부터 [-3]인 뒤에서 세 번째 값까지 출력하면 된다. 그러므로 10, 20, 30이 답으로 출력된다.        
[16] **풀이** : b[-2][1] 중에서 b[-2]는 뒤에서 두 번째 값인 "HCit"이다. b[-2]의 [1]은 "HCit"의 인덱스1 값인 C가 출력된다.         
[17] **풀이** : a[::-2]는 처음부터 끝까지 출력하되, 인덱스 짝수번째 값을 뒤에서부터 출력하므로 60, 40, 20이 출력된다.

---

In [23]:
# [18-19]
# 내포(Comprehension) 표기
# 리스트 내포 표기

monk_fish_team = [158, 157, 163, 157, 145]

total = sum(monk_fish_team)    # 리스트의 합계
length = len(monk_fish_team)   # 리스트의 요소 수(길이)
mean = total/length            # 평균을 구함

# 리스트의 분산(variance)를 구하는 for-loop 코드
# variance = 0
# for height in monk_fish_team:
#     variance = variance+(height-mean)**2
#
# variance = variance/length

# 리스트 내포 표기로 분산을 계산하는 식
# [18] 적절한 함수?  , [19] 분산을 계산하기 위한 변수는? 

variance = sum([(h - mean)**2 for h in monk_fish_team])/length

variance

# [18.보기] abs, mean, add, sum, sqrt, var

35.2

[18] **풀이** : variance는 낚시팀 팀원 각각의 (키-평균)의 제곱을 한 모든 값을 합하여 인원수로 나눈 값이므로, 팀원 각각의 결과를 합하는 함수로 sum이 들어가야 한다.    
[19] **풀이** : (height-mean)이므로 (h-___)에 들어갈 값은 mean이다.

---

In [24]:
# [20]
# lambda 식

x=4
list(map(lambda z : x % z, range(2,x)))  #[20] 계산 결과는

[0, 1]

**풀이** : x값인 4를 lambda식에 넣으면 4를 나누는 값은 2와 3이 된다. 그 결과는 다음과 같다.      
 4/2 = 2 나머지 0               
 4/3 = 1 나머지 1    이므로 답은 0과 1이다.

---

In [0]:
#
# [Bonus] lambda 함수 정의

# function square(x)
def square(x): 
    return x*x

# square() 함수를 lambda 식으로 정의하시오. (Bonus +1)

square = lambda x:x*x

**풀이** : x 값을 받으면 x*x를 계산한 값을 출력하게 된다.

In [30]:
print(square(1))
print(square(2))
print(square(3))
print(square(4))
print(square(5))

1
4
9
16
25


In [0]:
# 위는 x값이 들어왔을 때 그 값을 제곱한 결과이다.